support an arbitrary SecurityContext block for the main deployment

nefischer opened the pull request:
Signed-off-by: Nicolas Fischer <nicolas@emberspark.io>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**: allow any parameters defined in the core kubernetes SecurityContext spec to be specified as part of the `securityContext` parameter

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #1701 

**Special notes for your reviewer**:

[security-context-deprecation.md](./test/fixtures/security-context-deprecation.md) explains the changes made

[test-securitycontext-deprecation.sh](./hack/test-securitycontext-deprecation.sh) shows the tests run, including cases that test backwards compatibility.
 
**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Support for additional `securityContext` parameters.
All parameters defined in the core API spec for [SecurityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#securitycontext-v1-core) are now supported if the `securityContext.enabled` parameter is not specified.

Migration guide:

Until now, if `securityContext.enabled` was set to `true`, `securityContext.runAsUser` and `securityContext.fsGroup` would default to `1001`.
From now on, if `securityContext.enabled` is not specified, `securityContext.runAsUser` and `securityContext.fsGroup` have to be defined explicitly, as do any additional parameters from the [SecurityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#securitycontext-v1-core) specification.

Examples:

Example 1:
Old `values.yaml`:
securityContext:
  enabled: true

New `values.yaml`:
securityContext:
  fsGroup: 1001
  runAsUser: 1001

Example 2:
Old `values.yaml`:
securityContext:
  enabled: true
  fsGroup: 1111

New `values.yaml`:
securityContext:
  fsGroup: 1111
  runAsUser: 1001

Example 3:
Old `values.yaml`:
securityContext:
  enabled: false

New `values.yaml`:
securityContext: {}

To specify additional parameters, simply specify them in the yaml block, e.g.:
securityContext:
  fsGroup: 1111
  runAsUser: 2222
  allowPrivilegeEscalation: false
  readOnlyRootFilesystem: true
  privileged: false
```

munnerz commented on test/fixtures/security-context-deprecation.md:
```diff
@@ -0,0 +1,178 @@
+# Logic for deprecating securityContext.enabled
```
I think it'd be best to include this file under the 'design' directory - it'll be a helpful reference to link people to, and less buried! 😄

Feel free to copy it over as-is (without changing the content).

You should include the current date in the filename, e.g. `20191217-filename.md`.

nefischer replied:
File copied as requested 😄

munnerz replied:
Can you also remove the file from this directory? I should have said 'move it over' 😅

nefischer replied:
No worries. Done.

nefischer commented:
Hey @munnerz is there anything else I need to do (I copied the markdown file to the design folder as requested)?

munnerz commented:
/ok-to-test

nefischer commented:
Have fixed the linting issue in values.yaml, but not sure why the tests are timing out as it seems unrelated to what I changed.
Also rebased on top of master because that branch had moved on since the PR was raised.

munnerz commented:
Thanks! 😄 

/lgtm
/approve
